ROS2 Gem bare bones

adamdbrw opened the pull request:
First (barely) usable version of ROS2 Gem.
- ros2 target cmake function
- compilation flag: enable exceptions
- simulation clock. Time is published to /clock topic.
- spinner, executor in system component
- simple preview icon

Note that more Gem code is ready for PRs once this one is accepted.

Issues affected: 
1. Fixes #8484 (a consideration for frequency and threading can be separated out)
2. Adds progress to #8457. What will be left is to add some dependency checks such as ros2 version as well as enable Gem cmake use / include in Projects.

lumberyard-employee-dm commented on Gems/ROS2/Code/Platform/Common/Clang/ros2_editor_static_clang.cmake:
```diff
@@ -0,0 +1,11 @@
+# SPDX-License-Identifier: Apache-2.0 OR MIT
+#
+
+set(LY_COMPILE_OPTIONS
+        PRIVATE
+            -fexceptions
+)
```
You will need to include this compiler specific file in the CMakeLists.txt PLATFOMR_INCLUDE section for the `ROS2.Editor` target in order to get it to be used.

```
PLATFORM_INCLUDE_FILES
        Platform/Common/${PAL_TRAIT_COMPILER_ID}/ros2_editor_static_${PAL_TRAIT_COMPILER_ID_LOWERCASE}.cmake
```

lumberyard-employee-dm commented on Gems/ROS2/Code/Source/ROS2Module.cpp:
```diff
@@ -0,0 +1,17 @@
+
```
Some of these files are missing copyright helpers
1.Gems/ROS2/Code/Source/ROS2EditorSystemComponent.h
2. Gems/ROS2/Code/Source/ROS2Module.cpp
3. Gems/ROS2/Code/Source/ROS2EditorSystemComponent.cpp
4. Gems/ROS2/Code/Source/ROS2EditorModule.cpp
5. Gems/ROS2/Code/CMakeLists.txt
6. Gems/ROS2/CMakeLists.txt

adamdbrw replied:
I will add them. Would you say it also makes sense to also add them to the Gem template?

adamdbrw replied:
yes, though it is off by default:
```
  -kl, --keep-license-text
                        Should license in the template files text be kept in
                        the instantiation, default is False, so will not keep
                        license text by default. License text is defined as
                        all lines of text starting on a line with
                        {BEGIN_LICENSE} and ending line {END_LICENSE}.
```

lumberyard-employee-dm replied:
Then that sounds like we have a bug with some of the files then applying the copyright, when they shouldn't.

adamdbrw replied:
It was my omission - I was not aware about the flag previously, included license manually for some but not all files.

lumberyard-employee-dm commented on Gems/ROS2/Code/CMakeLists.txt:
```diff
@@ -0,0 +1,167 @@
+    FILES_CMAKE
+        ros2_files.cmake
+        ${pal_dir}/ros2_${PAL_PLATFORM_NAME_LOWERCASE}_files.cmake
+    PLATFORM_INCLUDE_FILES
+        ${CMAKE_CURRENT_LIST_DIR}/Platform/Common/${PAL_TRAIT_COMPILER_ID}/ros2_editor_static_${PAL_TRAIT_COMPILER_ID_LOWERCASE}.cmake
```
Did you mean to add this to the `ROS2.Editor.Static` library target instead of the Client `ROS2.Static` target?

lumberyard-employee-dm commented:
> @lumberyard-employee-dm I applied your review comments, thank you

There are build failures building the ROS2 Gem in Automated Review(AR): https://jenkins.build.o3de.org/job/O3DE/job/PR-8715/

```
12:46:10  WARNING,To build ROS2 Gem a ROS distribution needs to be sourced, but none detected
12:46:10  WARNING, missing from env. ROS 2 needs to be sourced to build ROS2 Gem
12:46:10  CMake Error at Gems/ROS2/Code/ros2_target_depends.cmake:27 (find_package):
12:46:10    By not providing "Findrclcpp.cmake" in CMAKE_MODULE_PATH this project has
12:46:10    asked CMake to find a package configuration file provided by "rclcpp", but
12:46:10    CMake did not find one.
12:46:10  
12:46:10    Could not find a package configuration file provided by "rclcpp" with any
12:46:10    of the following names:
12:46:10  
12:46:10      rclcppConfig.cmake
12:46:10      rclcpp-config.cmake
12:46:10  
12:46:10    Add the installation prefix of "rclcpp" to CMAKE_PREFIX_PATH or set
12:46:10    "rclcpp_DIR" to a directory containing one of the above files.  If "rclcpp"
12:46:10    provides a separate development package or SDK, be sure it has been
12:46:10    installed.
12:46:10  Call Stack (most recent call first):
12:46:10    Gems/ROS2/Code/CMakeLists.txt:41 (target_depends_on_ros2_packages)
12:46:10  
12:46:10  
12:46:10  -- Configuring incomplete, errors occurred!
```

It looks like the Gem needs to bundle the Findrclcpp.cmake script and add it to the 3rdParty search path in order to find the rclcpp library

adamdbrw commented:
> There are build failures building the ROS2 Gem in Automated Review(AR)

This is not that simple. Please take a look at issue #8457.

ROS2 has its own way of ensuring packages visibility, versioning and ENV population. This is done through workspace overlays enabled by [environment sourcing](https://docs.ros.org/en/rolling/Tutorials/Configuring-ROS2-Environment.html). One of benefits is an ability to switch versions / implementations readily. The choice of version and which packages are visible is a conscious action by the user. Find*.cmake goes against this principle.

What I proposed before is to have it work the ROS 2 way. To use the ROS2 Gem, the user needs to:
1. Have ros2 installed (and any number of ros2 workspaces built)
2. Source their ros2 installation and any number of overlays.
3. Build / run ROS2 Gem project.

For AR, this could mean some special handling (dependency plus a custom command?).

Let me know what you think.

lumberyard-employee-dm commented:
> > There are build failures building the ROS2 Gem in Automated Review(AR)
> 
> Unfortunately, this is not that simple. Please take a look at issue #8457.
> 
> ROS2 has its own way of ensuring packages visibility, versioning and ENV population. This is done through workspace overlays enabled by [environment sourcing](https://docs.ros.org/en/rolling/Tutorials/Configuring-ROS2-Environment.html). One of benefits is an ability to switch versions / implementations readily. The choice of version and which packages are visible is a conscious action by the user. Find*.cmake goes against this principle.
> 
> What I proposed before is to have it work the ROS 2 way. To use the ROS2 Gem, the user needs to:
> 
>     1. Have ros2 installed: [ROS2 installation guide for Galactic](https://docs.ros.org/en/galactic/Installation/Ubuntu-Install-Debians.html#installing-ros-2-via-debian-packages). Selecting `ros-galactic-ros-base` should be enough.
> 
>     2. Source their ros2 installation and any number of overlays.
> 
>     3. Build / run ROS2 Gem project.
> 
> 
> For AR, this could mean some special handling (dependency plus a custom command?). Is there any way I could help with that?
> 
> The warnings that you see:
> 
> > 12:46:10  WARNING,To build ROS2 Gem a ROS distribution needs to be sourced, but none detected
> > 12:46:10  WARNING, missing from env. ROS 2 needs to be sourced to build ROS2 Gem
> 
> Should probably be escalated to errors.
> 
> Let me know what you think.

Well a Gem on it's own shouldn't cause a configuration or build failure.

One thing that could be done is to have the ROS2 detect if ros2 installed in it's CMakeLists.txt.
If the ROS2 is not setup the CMakeLists for the Gem could early return at that point, as well as log warning message about the Gem needing to have ROS2 installed.

We do something similar with the Wwise SDK and the AudioEngineWwise Gem: https://github.com/o3de/o3de/blob/development/Gems/AudioEngineWwise/Code/CMakeLists.txt#L9-L28

Do you have any insights on this @amzn-phist?

adamdbrw commented:
@lumberyard-employee-dm review changes applied

adamdbrw commented:
I see, plenty of copyright headers missing, especially from platform-specific files. I will fix that.

lumberyard-employee-dm commented:
It looks like the copyright validator has failed: https://jenkins.build.o3de.org/blue/organizations/jenkins/O3DE/detail/PR-8715/2/pipeline

When you fix up those copyright comments, let me know and I can kick off another build.

amzn-phist commented:
Looks like it was updated already to work similar to the Wwise Gem.  The `Find<package>.cmake` script will determine whether the external SDK is valid and sets a `<Package>_FOUND` variable to indicate this.  Then the integration Gem will look at this variable to determine whether to declare targets or not.

One aspect of this setup that, I think, the Wwise Gem is missing is some basic messaging.  If someone has enabled the integration Gem and the SDK isn't found, it's basically silent to the user.

adamdbrw commented:
Things should validate now - I also removed empty or effectively empty platform-specific files.

amzn-sj commented on Gems/ROS2/Code/Source/ROS2SystemComponent.h:
```diff
@@ -0,0 +1,66 @@
+        ////////////////////////////////////////////////////////////////////////
+    
+    private:
+        std::shared_ptr<rclcpp::Node> m_ros2Node;
+        std::shared_ptr<rclcpp::executors::SingleThreadedExecutor> m_executor;
```
Nit: `m_executor` could be an AZStd::shared_ptr. 
I see that `m_ros2Node` is being passed to `add_node` so we don't have an option other than `std::shared_ptr`.

lumberyard-employee-dm commented on Gems/ROS2/Code/Source/ROS2EditorSystemComponent.h:
```diff
@@ -0,0 +1,39 @@
+    /// System component for ROS2 editor
+    class ROS2EditorSystemComponent
+        : public ROS2SystemComponent
+        , private AzToolsFramework::EditorEvents::Bus::Handler
```
The EditorEvents::Bus isn't actually being handled by this Editor component, so it can be removed.

It looks like the Gem Template adds this EBus handler to the EditorComponent, but doesn't actually implement the Event functions.

lumberyard-employee-dm reviewed (approved):
The change has passed AR: https://jenkins.build.o3de.org/job/O3DE/job/PR-8715/
It can be submitted to the repo now, but I want to make sure the @o3de/sig-core group gets a chance to review the RFC https://github.com/o3de/sig-core/issues/33

amzn-sj reviewed (commented):
I started another AR run after the last commit.

spham-amzn commented on Gems/ROS2/Code/Source/Clock/SimulationClock.cpp:
```diff
@@ -0,0 +1,53 @@
+        }
+        else
+        {
+            AZ_Warning("ROS2SystemComponent", false, "No ITime interface available");
```
Clarification: Is this an error or a warning? If the elapsed time is critical and is used by other parts of the system, this should be an error, unless downstream subscribers know how to handle 'zero' elapsed times

adamdbrw replied:
This would be an error for the simulation, since ros2 packages would not be able to run many time-sensitive algorithms (e.g. sensor fusion). I escalated it in the new changeset.

spham-amzn commented on Gems/ROS2/Code/Source/ROS2EditorSystemComponent.cpp:
```diff
@@ -0,0 +1,58 @@
+        }
+    }
+
+    ROS2EditorSystemComponent::ROS2EditorSystemComponent() = default;
+
+    ROS2EditorSystemComponent::~ROS2EditorSystemComponent() = default;
```
minor: You can probably just declare the default constructor/destructor in the header file only

lumberyard-employee-dm replied:
This is part of Gem template.
Also defining the default in the .cpp file allows the gem maintainers to update the System Components constructor/destructor without a potential longer compile if the system component header is included in multiple files

spham-amzn replied:
Ok, that makes sense

spham-amzn commented on Gems/ROS2/Code/Source/ROS2EditorSystemComponent.h:
```diff
@@ -0,0 +1,38 @@
+        AZ_COMPONENT(ROS2EditorSystemComponent, "{34fa5d9a-956b-4655-a6bc-1d57dce8e7a2}", BaseSystemComponent);
+        static void Reflect(AZ::ReflectContext* context);
+
+        ROS2EditorSystemComponent();
+        ~ROS2EditorSystemComponent();
```
(minor comment above)
```c++
ROS2EditorSystemComponent() = default;
~ROS2EditorSystemComponent() = default;
```

spham-amzn commented on Gems/ROS2/Code/ros2_target_depends.cmake:
```diff
@@ -0,0 +1,35 @@
+    if (NOT ROS2_FOUND)
+        return()
+    endif()
+    #message("Building ROS2 Gem with ros2 $ENV{ROS_DISTRO}")
```
You either remove this line if not needed, or enable it and print in  'DEBUG' mode
```
message(DEBUG "Building ROS2 Gem with ros2 $ENV{ROS_DISTRO}")
```

spham-amzn commented on Gems/ROS2/Code/ros2_target_depends.cmake:
```diff
@@ -0,0 +1,35 @@
+        list(APPEND _ros2_include_directories "${_ros2_packages_path}/include")
+    endforeach()
+    foreach(_package IN LISTS ARGN)
+        #message("processing package: ${_package}")
```
Same comment as above

amzn-sj commented on Gems/ROS2/Code/CMakeLists.txt:
```diff
@@ -0,0 +1,158 @@
+# Add the ROS2.Static target
+# Note: We include the common files and the platform specific files which are set in ros2_common_files.cmake
+# and in ${pal_dir}/ros2_${PAL_PLATFORM_NAME_LOWERCASE}_files.cmake
+ly_add_target(
+    NAME ROS2.Static STATIC
+    NAMESPACE Gem
+    FILES_CMAKE
+        ros2_files.cmake
+    INCLUDE_DIRECTORIES
+        PUBLIC
+            Include
```
Currently, `ROS2Bus.h` is being included as part of the STATIC target ROS2.Static. So if a project(or any other Gem) wants to use the `ROS2RequestBus` or the `ROS2Interface`, they would need to add either `ROS2.Static` or `ROS2` module(since that also adds Include as a public include dir) as a build dependency which would result in linking those unnecessarily. If they link against `ROS2` module instead of static, it's probably not a big deal but it could be avoided.
There's also the issue of installed(SDK) version of the engine [propagating private build dependencies as runtime dependencies](https://github.com/o3de/o3de/pull/5581). So if some other gem `MyGem` has a private build dependency on `ROS2`, and the project has both `MyGem` and `ROS2` gem enabled, we would end up trying to load both `ROS2.Editor` and `ROS2` modules as runtime dependencies which would lead to a bug similar to [this](https://github.com/o3de/o3de/issues/5822). This is probably an unlikely scenario but it's something to consider.
To avoid this, we could move the `ROS2Bus.h` to a header only target called `ROS2.API` and any gem/project that needs access can just add that as a build dependency. We do something similar for [ScriptCanvas](https://github.com/o3de/o3de/pull/5923) and [LmbrCentral](https://github.com/o3de/o3de/pull/6012) gems.
Not sure if this is relevant to this gem since I don't know how it will end up being used. But I thought it might be useful to consider since we've run into problems like this before.

adamdbrw replied:
I have added such API target in the most recent changeset

adamdbrw commented:
I have addressed the review comments. The AR seems to have failed recently but it seems unrelated to the PR (`AutomatedTesting::PhysicsTests_Main.main::TEST_RUN (Failed)`)

spham-amzn commented on Gems/ROS2/Code/ros2_target_depends.cmake:
```diff
@@ -18,18 +18,15 @@ function(target_depends_on_ros2_packages TARGET_NAME)
     endforeach()
     foreach(_package IN LISTS ARGN)
-        #message("processing package: ${_package}")
+        message("DEBUG processing package: ${_package}")
```
should be:
```
message(DEBUG "processing package: ${_package}")
```

adamdbrw commented:
Closing this PR since it was decided by sig-core that the Gem will reside in its own repository (https://github.com/RobotecAI/o3de-ros2-gem).
